### ISSUE TITLE
Icons aren't displaying properly

### ISSUE THREAD

**Issue by kamui-fin** [gokcehan/lf#583]
I used the wiki article for adding icons to lf through the `LF_ICONS` env variable. I also have nerd fonts already installed, yet, it doesn't seem to be displaying properly. This screenshot is with the `st` terminal but i've also tried it on the kitty terminal and it still doesn't display the icons properly.

![image](https://user-images.githubusercontent.com/65237048/108282255-e810da00-7146-11eb-9b7a-8a24981c23b4.png)

Is this a bug or could I have configured something wrong?

**Comment by gokcehan** [comment]
@kamui-7 This is not intended behavior. Can you try this without setting any icons? (e.g. `LF_ICONS= lf`) If it's working, can you try with a minimal example? (e.g. `LF_ICONS='di=:*.c=' lf`). Also what shell are you using? This might also be an encoding issue. Are you using utf-8 in your terminal? Also what are your language variables? (e.g. LC_ALL, LC_CTYPE, and LANG)

**Comment by cubercsl** [comment]
The same problem.

`LANG=en_US.UTF-8 LF_ICONS='di=:*.c=' lf`

<img width="455" alt="图片" src="https://user-images.githubusercontent.com/22931465/133645667-5dd59eff-4d08-46b5-9f51-93a739153814.png">

`LANG=zh_CN.UTF-8 LF_ICONS='di=:*.c=' lf`

<img width="346" alt="图片" src="https://user-images.githubusercontent.com/22931465/133645936-6f9d5703-b8d9-430e-a0ee-c0b783a218e1.png">

**Comment by ndav1d** [comment]
what shell are you using in `lf`? it it's `zsh` make sure to check out [this](https://github.com/gokcehan/lf/wiki/Tips#split-words-by-default-in-zsh)

**Comment by cubercsl** [comment]
> what `shell` are you using in `lf`? if it is `zsh` make sure to check [this](https://github.com/gokcehan/lf/wiki/Tips#split-words-by-default-in-zsh) out

I am using `fish` in `alacritty`.

**Comment by gokcehan** [comment]
@GenjiFujimoto I can produce similar glitches with tcell [hello world](https://github.com/gdamore/tcell/blob/master/_demos/hello_world.go) example with different locales so I suspect this is a tcell issue.

**Comment by pezcore** [comment]
I am seeing some, but not all, icons display incorrectly. Namely, the directory icon and the `jpeg` icon:

![lf](https://user-images.githubusercontent.com/8358974/181385246-25b85988-70a8-466d-8d09-5ba121c81943.png)

I am seeing this in both kitty and alacritty. I believe this is related to `lf` because when I cat the `icons` file the directory icon displays correctly in the same terminal:

![icons_file](https://user-images.githubusercontent.com/8358974/181385496-7ddbafe6-da15-4ce6-b843-b393c670b2bd.png)

**Comment by copypasteonly** [comment]
> 

Hello have you found the solution to this? I've the same setup as yours except that my font is Codelia patched with NerdFont, my directory icon is the one that's not working.

**Comment by ilyagr** [comment]
I've investigated this a bit, but found no solution. It seems that the reason for this behavior is that the symbol `lf`'s `icons.example` uses to show directories is not a valid Unicode symbol. Only some hacked fonts (e.g. Nerdist) show it as a folder. Other fonts show it as the J-like squiggle shown on @pezcore's screenshot. (Most fonts show nothing at all).

In my case, also on Kitty, I've noticed that if I paste the folder symbol into Vim and move the cursor over it, it turns into the J-like squiggle.

You can try to find out why Kitty falls back to another font and how to force Kitty to use Nerd fonts everywhere. In any case, this seems like a Kitty problem, not an lf problem.

**Comment by mymmrac** [comment]
@kamui-fin this can be because you are using Nerd Font Mono instead of Nerf Font (regular)

For example, JetBrainsMono Nerd Font will be displayed properly but JetBrainsMono Nerd Font Mono will be the same as for you (at least on my machine)

Try to change your font, maybe this will help

**Comment by DusanLesan** [comment]
Maybe you can try using custom icon definitions. I have not even used icons bit to me it looks straightforward. I have seen implementation in Luke Smith's dotfiles.

**Comment by PierrotAWB** [comment]
@cubercsl Sorry for the bump, but can you post the output of `grep -v ^# /etc/locale.gen`? I had this issue earlier and setting `LANG` was insufficient. Uncommenting the relevant line (`en_US.UTF-8 UTF-8`) in the foregoing file fixed my issue. Seemed to have something to do with system locale.

**Comment by MagnesiaReal** [comment]
> I am seeing some, but not all, icons display incorrectly. Namely, the directory icon and the `jpeg` icon:
> 
> ![lf](https://user-images.githubusercontent.com/8358974/181385246-25b85988-70a8-466d-8d09-5ba121c81943.png)
> 
> I am seeing this in both kitty and alacritty. I believe this is related to `lf` because when I cat the `icons` file the directory icon displays correctly in the same terminal:
> 
> ![icons_file](https://user-images.githubusercontent.com/8358974/181385496-7ddbafe6-da15-4ce6-b843-b393c670b2bd.png)
> 
> Also, if it helps, My terminal it `kitty`, my shell is `zsh 5.9`, my font is `JetBrains Mono` and my `LANG` is `en_US.utf8`. My current config correctly display Nerdfont icons in every other application except `lf`
> 
> My `lf` version is `r27`

I found maybe a hacked solution

go to your kitty.conf and set:
`symbol_map U+1000-U+ffee JetBrainsMono Nerd Font`
that's all

**Comment by coma409** [comment]
> @GenjiFujimoto I can produce similar glitches with tcell [hello world](https://github.com/gdamore/tcell/blob/master/_demos/hello_world.go) example with different locales so I suspect this is a tcell issue.

Thanks for the hint ,I think it is the tcell's issue.

https://github.com/microsoft/terminal/issues/370#issue-408548048
https://github.com/gdamore/tcell/issues/531#issue-1240220589

In CJK environment ,Ambiguous width character will be glyphs to a different width.

I fix it by set the envirement `export RUNEWIDTH_EASTASIAN=0` in my `$HOME/.xprofile`.

```
export LANG="zh_CN.UTF-8"
export LC_ALL="zh_CN.UTF-8"

export GDK_SCALE=2
export GDK_DPI_SCALE=0.5
export QT_AUTO_SCREEN_SCALE_FACTOR=

export GTK_IM_MODULE=fcitx
export QT_IM_MODULE=fcitx
export XMODIFIERS=@im=fcitx

export GOROOT=/usr/lib/go
export GOPATH=$HOME/go
export GOBIN=$HOME/go/bin
export PATH=$PATH:$GOROOT/bin:$GOBIN
export GO111MODULE=on

export RUNEWIDTH_EASTASIAN=0
```

![lf_1](https://user-images.githubusercontent.com/51706716/219574900-377d6bcc-1c12-47af-bb5b-119bd3e514a7.png)

OS: `ArchLinux x86_64`
Locale: `LANG=zh_CN.UTF-8`
Terminal Emulator: `Alacritty`

**Comment by pingarelho** [comment]
> I am seeing some, but not all, icons display incorrectly. Namely, the directory icon and the `jpeg` icon:
> 
> ![lf](https://user-images.githubusercontent.com/8358974/181385246-25b85988-70a8-466d-8d09-5ba121c81943.png)
> 
> I am seeing this in both kitty and alacritty. I believe this is related to `lf` because when I cat the `icons` file the directory icon displays correctly in the same terminal:
> 
> ![icons_file](https://user-images.githubusercontent.com/8358974/181385496-7ddbafe6-da15-4ce6-b843-b393c670b2bd.png)
> 
> Also, if it helps, My terminal it `kitty`, my shell is `zsh 5.9`, my font is `JetBrains Mono` and my `LANG` is `en_US.utf8`. My current config correctly display Nerdfont icons in every other application except `lf`
> 
> My `lf` version is `r27`

currently having this exact issue

**Comment by promitheas17j** [comment]
Experiencing this issue as well. Currently only noticed with jpeg files. Instead of the picture icon it displays a rocket icon.

![2023-11-29_00-18](https://github.com/gokcehan/lf/assets/12212713/b770b2f3-670b-455a-955b-4f7d5a622757)
![2023-11-29_00-25](https://github.com/gokcehan/lf/assets/12212713/8cd54432-3327-4230-8279-89f810df1ab6)

What I did is copy-paste a lot of the contents of the [example file](https://github.com/gokcehan/lf/blob/master/etc/icons.example) into my icons file in the lf directory. The I added the line ```set icons``` in my lfrc.

**Comment by BluishHumility** [comment]
The terminal needs to be configured to use a nerd font to properly show the glyphs in the example file. You can either configure a nerd font as the primary font of the terminal, or you can install `ttf-nerd-fonts-symbols` and configure the terminal to use that as a fallback font.

Having a nerd font installed is not enough; it is possible to have a nerd font installed but configure your terminal to use the "regular" version of the font, in which case the icons will not show up correctly.

When you are configuring the font, check `fc-list` to be sure you are getting the spelling and capitalization of the font exactly right when you add it to your terminal config. If you don't type it exactly right it won't work. Check `fc-list | grep -i nerd` to see what your options are for nerd fonts if you aren't sure. 

Using Foot terminal as an example, if I set my font like this, it displays the glyphs correctly:

```
font=JetBrainsMonoNerdFont-Regular:size=13
```

Using a non-nerd font for the primary terminal font and configuring the symbols font as a fallback also works.

```
font=JetBrainsMono-Regular:size=13,SymbolsNerdFont-Regular:size=13
```

Setting the font to just `JetBrainsMono-Regular` does not work. The glyphs will be displayed as they are in the various screenshots in this thread.

**Comment by CatsDeservePets** [comment]
As @BluishHumility said, this seems to have nothing to do with `lf`. I will close this issue.
If anyone believes otherwise, feel free to comment on this issue again and I will re-open it.

**Comment by promitheas17j** [comment]
Not sure if I really am the one with the messed up config, but my alacritty font config is this:

```
[font.normal]
family = "CaskaydiaCoveNerdFont"
style = "Regular"
```

And yet lf still shows the jpeg icons as rockets. Keep in mind nerdfont icons work just fine in my other TUI programs, such as nvim's bottom status bar, or in the file explorer I included within nvim.

It also appears in:

```
$ fc-list | grep CaskaydiaCoveNerdFont
// other fonts with that in their name
/usr/share/fonts/TTF/CaskaydiaCoveNerdFont-Regular.ttf: CaskaydiaCove Nerd Font,CaskaydiaCove NF:style=Regular
// other fonts with that in their name
```

If its something I misconfigured I'm sorry for recalling attention to this issue. Otherwise I'd be happy to help track the issue down.

**Comment by CatsDeservePets** [comment]
The rocket icon is used for executable files. Any chance you are using `wsl` (Windows subsystem Linux)? There seem to be problems with files being marked as executable under `wsl` (not a problem with `lf`).

**Comment by promitheas17j** [comment]
No, im on native Arch Linux, however I just checked and Ive got the following installed through my package manager:

```
$ pacman -Qi ttf-cascadia-code-nerd
Name            : ttf-cascadia-code-nerd
Version         : 3.4.0-1
Description     : Patched font Cascadia Code (Caskaydia) from nerd fonts library
Architecture    : any
URL             : https://github.com/ryanoasis/nerd-fonts
Licenses        : OFL-1.1-RFN
Groups          : nerd-fonts
Provides        : ttf-font-nerd
Depends On      : None
Optional Deps   : None
Required By     : nerd-fonts-meta
Optional For    : None
Conflicts With  : nerd-fonts-cascadia-code
Replaces        : nerd-fonts-cascadia-code  otf-cascadia-code-nerd
Installed Size  : 96.70 MiB
Packager        : T.J. Townsend <blakkheim@archlinux.org>
Build Date      : Thu 24 Apr 2025 21:27:28 EEST
Install Date    : Tue 29 Apr 2025 20:46:01 EEST
Install Reason  : Installed as a dependency for another package
Install Script  : No
Validated By    : Signature
```

Digging around in the upstream URL it seems ryanoasis included the microsoft version of this particular nerdfont. Perhaps its related to why there is an issue on wsl as well as on my setup. Thanks for mentioning the WSl issue, I wouldn't have thought to look into it if you hadn't

**Comment by joelim-work** [comment]
The screenshot from https://github.com/gokcehan/lf/issues/583#issuecomment-1830857909 shows that the file is executable. The stat info on the bottom left corner says `-rwxrwxrwx`, so the expected icon is a rocket as filetypes take precedence over executables, see #992.

**Comment by promitheas17j** [comment]
> The screenshot from [#583 (comment)](https://github.com/gokcehan/lf/issues/583#issuecomment-1830857909) shows that the file is executable. The stat info on the bottom left corner says `-rwxrwxrwx`, so the expected icon is a rocket as filetypes take precedence over executables, see [#992](https://github.com/gokcehan/lf/issues/992).

Youre absolutely right. I suspect that this is because those particular images were probably taken off an NTFS drive where windows set them as executable and thats why they have those permissions but others don't. I simply removed the executable permission from them and now they display correctly, so it also doesn't seem to be an issue with the particular nerdfont I use either.

I suppose this solves the mystery, and this can be closed once again. Thank you :)

**Comment by joelim-work** [comment]
No worries.

I think the original issue is possibly caused by locales and might be related to #1159. Will close this then.
